File: ui/src/main/java/com/xmartlabs/bigbang/ui/common/recyclerview/RecycleItemType.java
package com.xmartlabs.bigbang.ui.common.recyclerview;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * A recycler view item, which relate an item {@link I} with a {@link android.support.v7.widget.RecyclerView.ViewHolder}
 * {@link VH}.
 *
 * @param <I>  Item type.
 * @param <VH> View holder type.
 */
public interface RecycleItemType<I, VH extends RecyclerView.ViewHolder> {
  /**
   * Its called to crate a new view holder of the type {@link VH}.
   *
   * @param parent the view parent of the view holder.
   * @return A {@link android.support.v7.widget.RecyclerView.ViewHolder} of the type {@link VH}.
   */
  @NonNull
  VH onCreateViewHolder(@NonNull ViewGroup parent);
  /**
   * It's called to bind the {@link android.support.v7.widget.RecyclerView.ViewHolder} {@link VH} with the item of the
   * type {@link I}
   *
   * @param viewHolder {@link android.support.v7.widget.RecyclerView.ViewHolder} to be bound.
   * @param item       Item to be to be bound.
   * @param position   Position in the adapter.
   */
  void onBindViewHolder(@NonNull VH viewHolder, @NonNull I item, int position);
}

File: ui/src/main/java/com/xmartlabs/bigbang/ui/common/recyclerview/RecycleItemType.kt
package com.xmartlabs.bigbang.ui.common.recyclerview

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

/**
 * A recycler view item, which relates an item [I] with a [android.support.v7.widget.RecyclerView.ViewHolder].
 *
 * @param <I>  Item type.
 * *
 * @param <VH> View holder type.
 */
interface RecycleItemType<in I : Any, VH : RecyclerView.ViewHolder> {
  /**
   * Called to crate a new view holder of the type [VH].
   *
   * @param parent the parent view of the view holder.
   * *
   * @return A [android.support.v7.widget.RecyclerView.ViewHolder] of the type [VH].
   */
  fun onCreateViewHolder(parent: ViewGroup): VH

  /**
   * Called to bind the [android.support.v7.widget.RecyclerView.ViewHolder] [VH] with the item of the ype [I]
   *
   * @param viewHolder [android.support.v7.widget.RecyclerView.ViewHolder] to be bound.
   * *
   * @param item       Item to be bound.
   * *
   * @param position   Position in the adapter.
   */
  fun onBindViewHolder(viewHolder: VH, item: I, position: Int)
}
